### ISSUE TITLE
yield + Gratuitous parentheses around expression

### ISSUE THREAD

**Issue by pgherveou** [eslint/eslint#1981]
Hey there thanks for the cool project,

I noticed the following error:
the following code produce  a "Gratuitous parentheses around expression"

```js
if ((yield someGenerator()) === 1) {
}
```

**Comment by nzakas** [comment]
Yup, the rule flags unnecessary parentheses. What is it you are reporting?

**Comment by michaelficarra** [comment]
@nzakas `yield` has the precedence of an `AssignmentExpression`. Those parentheses are necessary, otherwise the whole following expression would be the `yield` operand. I wrote this rule, and it probably doesn't work because it was written before we had `yield` support (and probably before that was even in the ES6 draft!).

**Comment by nzakas** [comment]
Ah okay, thanks for explaining. We can get that fixed.

**Comment by pgherveou** [comment]
thks @michaelficarra you were faster than me! 
I tried that too but by default it generate a  'Expected literal to be on the right side'

```
if (1 === yield someGenerator() ) {
}
```

**Comment by nzakas** [comment]
Working on this.

**Comment by Setonix** [comment]
I have the same warning in a command like this : 
Because of the additional parentheses I get the warning, however when I leave them out, ss will only contain 'either this'. What I want to get out is "Begin Line either this End of Line". So in that case the parentheses are needed. 

This seems like a silly example, but I use it in calling jscript libraries and writing the url for them.

`var t = 1;
var ss = 'Begin Line ' + ((t === 1)?  'either this' : 'or that') + ' End of Line';
alert(ss);
`

**Comment by michaelficarra** [comment]
You don't need to parenthesise `t === 1`.

**Comment by Setonix** [comment]
Aha, good to know, thanks
